pkg/vendir/fetch/hg/hg.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package hg
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	ctlconf "github.com/vmware-tanzu/carvel-vendir/pkg/vendir/config"
+	ctlfetch "github.com/vmware-tanzu/carvel-vendir/pkg/vendir/fetch"
+	ctlver "github.com/vmware-tanzu/carvel-vendir/pkg/vendir/versions/v1alpha1"
+)
+
+type Hg struct {
+	opts       ctlconf.DirectoryContentsHg
+	infoLog    io.Writer
+	refFetcher ctlfetch.RefFetcher
+}
+
+func NewHg(opts ctlconf.DirectoryContentsHg,
+	infoLog io.Writer, refFetcher ctlfetch.RefFetcher) *Hg {
+
+	return &Hg{opts, infoLog, refFetcher}
+}
+
+// nolint:golint
+type HgInfo struct {
+	SHA         string
+	Tags        []string
+	CommitTitle string
+}
+
+func (t *Hg) Retrieve(dstPath string, tempArea ctlfetch.TempArea) (HgInfo, error) {
+	if len(t.opts.URL) == 0 {
+		return HgInfo{}, fmt.Errorf("Expected non-empty URL")
+	}
+
+	err := t.fetch(dstPath, tempArea)
+	if err != nil {
+		return HgInfo{}, err
+	}
+
+	info := HgInfo{}
+
+	out, _, err := t.run([]string{"id", "-i"}, nil, dstPath)
+	if err != nil {
+		return HgInfo{}, err
+	}
+
+	info.SHA = strings.TrimSpace(out)
+
+	out, _, err = t.run([]string{"log", "-l", "1", "-T", "{desc|firstline|strip}", "-r" , info.SHA}, nil, dstPath)
+	if err != nil {
+		return HgInfo{}, err
+	}
+
+	info.CommitTitle = strings.TrimSpace(out)
+
+	return info, nil
+}
+
+func (t *Hg) fetch(dstPath string, tempArea ctlfetch.TempArea) error {
+	authOpts, err := t.getAuthOpts()
+	if err != nil {
+		return err
+	}
+
+	authDir, err := tempArea.NewTempDir("hg-auth")
+	fmt.Printf(">>> authDir: %s", authDir)
+	if err != nil {
+		return err
+	}
+
+	defer os.RemoveAll(authDir)
+
+	env := os.Environ()
+
+	hgURL := t.opts.URL
+
+	_, _, err = t.run([]string{"init"}, env, dstPath)
+	if err != nil {
+		return err
+	}
+
+	hgrcPath := filepath.Join(dstPath, ".hg", "hgrc")
```

Comment:
per https://github.com/vmware-tanzu/carvel-vendir/issues/88#issuecomment-891986018, we can drop that into separate location

Reply:
totally agree